### ISSUE TITLE
Iterations number stays at zero

### ISSUE THREAD

**Issue by tianBreznik** [scipy/scipy#9153]
Hello,

This is probably not a bug nor an issue, but rather a question about the possible reasons for the behaviour of the CG method of scipy.optimize.minimize.

I built a neural network with a single hidden layer for some simple pattern recognition, using the backpropagation algorithm for training. I first coded the network in Matlab, also using fmincg and the predicted results were 97% accurate. After this, I rewrote the code for the network in Python accounting for the differences between the languages and testing all the functions to be evaluated. 

My question is about the number of iterations and linear search steps taken by the algorithm. In Matlab, the algorithm iterated over the network a large number of times (depending on the maximum number of iterations I set) on the same training set as that which I used to train the network with the optimize.minimize algorithm. But for some reason I do not understand, the optimize.minimize algorithm iterates 0 number of times and only does 1 function and 1 gradient evaluation. I would be immensely grateful for any insight into this. 

Please understand that I am a first-year undergraduate student with very limited knowledge of machine learning and numerical analysis. I could not be more grateful for being able to use your product and I apologize in advance if this is a trivial problem.

The example given below is simply to show the cost function of the neural network and the backpropagation function which returns the gradient matrix. It is not illustrative of the neural network with the real dataset.

### Reproducing code example:
```

    import numpy as np
    from scipy.optimize import minimize
    
    
    A =  np.random.rand(10,10)
    labels = np.array([1, 0, 0, 1, 1, 1, 0, 1, 0, 1])
    lamb = 1
    #input size = 10, hidden size = 10
    W1 = np.random.randn(11, 10)
    #output size = 1
    W2 = np.random.randn(11 ,1)
    initial_nn_params = np.append(np.asarray(W1).reshape(-1), np.asarray(W2).reshape(-1))
    
    def sigmoid(self, s):
        # activation function
        return 1 / (1 + np.exp(-s))

    def sigmoidPrime(self, s):
        # derivative of sigmoid
        return s * (1 - s)

    def nnCostFunction(self, initialParams, inputInt, hiddenInt, outputInt, labelData, lambs):

        initial_weights_vec = initialParams
        inputSize = inputInt
        hiddenSize = hiddenInt
        outputSize = outputInt
        labels = labelData
        lambd_a = lambs

        firstWeights = np.reshape(initial_weights_vec[0:(hiddenSize * (inputSize + 1))], (hiddenSize, inputSize + 1))
        secondWeights = np.reshape(initial_weights_vec[hiddenSize * (inputSize +1):], (outputSize, hiddenSize + 1))
        m = len(A)

        J = 0

        foo_= np.ones((m, 1))
        a1 = np.hstack((foo_, A))
        z2 = np.dot(a1, firstWeights.T)
        a2 = sigmoid(z2)
        a2 = np.hstack((np.ones((len(a2),1)), a2))

        z3 = np.dot(a2, secondWeights.T)
        a3 = sigmoid(z3)
        hThetaX = a3


        for i in range(0, m):
            term1 = np.multiply(-labels[i, :], np.log(hThetaX[i, :]))
            term2 = np.multiply((np.ones((1, outputSize)) - labels[i, :]), np.log(np.ones((1, outputSize)) - hThetaX[i, :]))
            J = J + (term1 - term2).sum(axis=0)

        J = J / m

        regulator = (np.exp(firstWeights[:, 1:]).sum() + np.exp(secondWeights[:, 1:]).sum()) * (lambd_a / 2)
        J = J + regulator
        return J

    def nnCostDeriv(self, initialParams, inputInt, hiddenInt, outputInt, labelData, lambs):

        initial_weights_vec = initialParams
        inputSize = inputInt
        hiddenSize = hiddenInt
        outputSize = outputInt
        labels = labelData
        lambd_a = lambs

        firstWeights = np.reshape(initial_weights_vec[0:(hiddenSize * (inputSize + 1))], (hiddenSize, inputSize + 1))
        secondWeights = np.reshape(initial_weights_vec[hiddenSize * (inputSize + 1):], (outputSize, hiddenSize + 1))

        theta1_grad = np.zeros(np.shape(firstWeights))
        theta2_grad = np.zeros(np.shape(secondWeights))

        for j in range(0, m):
            a1 = np.hstack((1, A[i, :].T))
            z2 = np.dot(firstWeights, a1)
            a2 = np.hstack((1, self.sigmoid(z2)))

            z3 = np.dot(secondWeights, a2)
            a3 = sigmoid(z3)

            yy = ((1 == labels[j]).astype(int)).T

            delta_3 = a3 - yy
            delta_2 = np.multiply((np.dot(secondWeights.T, delta_3)), np.hstack((1, self.sigmoid(z2))))
            delta_2 = delta_2[1:]

            theta1_grad = theta1_grad + np.outer(delta_2, a1.T)
            theta2_grad = theta2_grad + np.outer(delta_3, a2.T)

        theta1_grad = (1 / m) * theta1_grad + (lambd_a / m) * np.hstack(
            (np.zeros((len(firstWeights),1)), firstWeights[:, 1:]))
        theta2_grad = (1 / m) * theta2_grad + (lambd_a / m) * np.hstack(
            (np.zeros((len(secondWeights),1)), secondWeights[:, 1:]))

        grad = np.append(np.asarray(theta1_grad).reshape(-1), np.asarray(theta2_grad).reshape(-1))
        return grad


    costFunction = lambda p: nnCostFunction(p, 10, 10, 1, labels, lamb)
    deriv = lambda der: nnCostDeriv(der, 10, 10, 1, labels, lamb)
    optimizationResults = optimize.minimize(costFunction, x0 = initial_nn_params, jac = deriv, options = {'disp':True,'maxiter':maxiter}, method="CG",)

```

### Error message:
This is not the error message, but the output received:
Optimization terminated successfully.
         Current function value: 9.027020
         Iterations: 0
         Function evaluations: 1
         Gradient evaluations: 1

### Scipy/Numpy/Python version information:
```
('1.1.0', '1.14.5', sys.version_info(major=2, minor=7, micro=10, releaselevel='final', serial=0))
```

**Comment by pv** [comment]
You can ask general usage questions e.g. on stackoverflow.com.

The code example you give is not runnable as-is. One possible reason for what you see may be that the gradient function calculates the gradient incorrectly. A second possible reason is that the problem is badly scaled, or the default tolerance is different (you do not give any tolerance specification to minimize).

**Comment by tianBreznik** [comment]
Hi, I have fixed the gradient function and now I am very confident that it is calculating the gradient correctly. I have also checked against the correct values from Matlab. But now I receive the following message at the end of the computation: "Warning: Desired error not necessarily achieved due to precision loss.". Any information on what could cause this would be very much appreciated!

**Comment by rgommers** [comment]
this still doesn't run, I stopped fixing it after the third failure. @tianBreznik your code should run when pasted in an empty file with `python filename.py`
